Communities encryption

vitvly opened the pull request:
Adds support for communities encryption in `protocol/encryption` package. Flow is this (taken from `protocol/encryption/README.md`:

1. `encryptor.GenerateCommunityKey()` generates a community key and stores it in in the DB.
There, 2 new tables are created: `communities_encryption` and `communities_encryption_cache`.
Each community key is uniquely identified by the `(communityId, keyId)` pair, where `keyId` is derived from a clock value.

2. `protocol.BuildCommunityKeyExchangeMessage` builds an 1-on-1 message containing the community key, given it's ID.

3. `protocol.BuildCommunityMessage` builds a community message with arbitrary payload, given `communityId`. It will use the latest community key available. `encryptor.encryptWithHR` encrypts the payload using Hash Ratchet algorithms. Intermediate hashes are stored in `communities_encryption_cache` table.

4. `protocol.HandleMessage` uses `encryptor.decryptWithHR` fn for decryption.

Commit:
Communities encryption

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (48)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 3b8a2e36 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/1/) | 2021-10-21 16:50:28 | ~50 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/1/consoleText) |
| :heavy_check_mark: | 3b8a2e36 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/1/) | 2021-10-21 16:53:03 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-211021-164941-3b8a2e.zip) |
| :heavy_check_mark: | 3b8a2e36 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/1/) | 2021-10-21 16:55:59 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-211021-164936-3b8a2e.aar) |
| | | | | | | |
| :x: | a756ad69 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/2/) | 2021-10-27 09:09:43 | ~21 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/2/consoleText) |
| :heavy_check_mark: | a756ad69 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/2/) | 2021-10-27 09:11:39 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-211027-090921-a756ad.zip) |
| :heavy_check_mark: | a756ad69 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/2/) | 2021-10-27 09:16:15 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-211027-090921-a756ad.aar) |
| | | | | | | |
| :x: | 3954e443 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/3/) | 2021-11-09 10:42:24 | ~28 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/3/consoleText) |
| :x: | 3954e443 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/3/) | 2021-11-09 10:45:59 | ~4 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/3/consoleText) |
| :x: | 3954e443 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/3/) | 2021-11-09 10:51:55 | ~10 min | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/3/consoleText) |
| | | | | | | |
| :x: | 8a50858d | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/4/) | 2021-11-10 11:41:22 | ~1 min | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/4/consoleText) |
| :x: | 8a50858d | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/4/) | 2021-11-10 11:42:49 | ~3 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/4/consoleText) |
| :x: | 8a50858d | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/4/) | 2021-11-10 11:44:24 | ~4 min | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/4/consoleText) |
| | | | | | | |
| :x: | 09ba49b2 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/5/) | 2021-11-11 13:42:46 | ~22 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/5/consoleText) |
| :x: | 09ba49b2 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/5/) | 2021-11-11 13:43:20 | ~55 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/5/consoleText) |
| :x: | 09ba49b2 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/5/) | 2021-11-11 13:43:57 | ~1 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/5/consoleText) |
| | | | | | | |
| :x: | 50cf1f80 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/6/) | 2021-11-12 11:22:12 | ~19 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/6/consoleText) |
| :x: | 50cf1f80 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/6/) | 2021-11-12 11:22:15 | ~23 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/6/consoleText) |
| :x: | 50cf1f80 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/6/) | 2021-11-12 11:23:16 | ~1 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/6/consoleText) |
| | | | | | | |
| :x: | fd470dba | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/7/) | 2021-11-12 12:13:06 | ~23 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/7/consoleText) |
| :x: | fd470dba | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/7/) | 2021-11-12 12:13:06 | ~23 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/7/consoleText) |
| :x: | fd470dba | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/7/) | 2021-11-12 12:14:06 | ~1 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/7/consoleText) |
| | | | | | | |
| :x: | 0a0d3ce5 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/8/) | 2021-11-12 12:26:44 | ~24 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/8/consoleText) |
| :x: | 0a0d3ce5 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/8/) | 2021-11-12 12:27:45 | ~1 min | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/8/consoleText) |
| :x: | 0a0d3ce5 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/8/) | 2021-11-12 12:27:47 | ~1 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/8/consoleText) |
| | | | | | | |
| :x: | 8fc6c126 | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/9/) | 2021-11-12 12:29:05 | ~17 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/9/consoleText) |
| :x: | 8fc6c126 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/9/) | 2021-11-12 12:29:05 | ~18 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/9/consoleText) |
| :heavy_check_mark: | 8fc6c126 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/9/) | 2021-11-12 12:34:15 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-211112-122842-8fc6c1.aar) |
| | | | | | | |
| :x: | 51a16a52 | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/10/) | 2021-11-17 10:02:20 | ~20 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/10/consoleText) |
| :x: | 51a16a52 | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/10/) | 2021-11-17 10:02:25 | ~24 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/10/consoleText) |
| :heavy_check_mark: | 51a16a52 | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/10/) | 2021-11-17 10:07:00 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-211117-100154-51a16a.aar) |
| | | | | | | |
| :heavy_check_mark: | 976f35e6 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/11/) | 2021-11-19 15:01:33 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-211119-145817-976f35-pr2412.zip) |
| :heavy_check_mark: | 976f35e6 | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/11/) | 2021-11-19 15:03:06 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-211119-145817-976f35-pr2412.zip) |
| :heavy_check_mark: | 976f35e6 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/11/) | 2021-11-19 15:06:46 | ~8 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-211119-145817-976f35-pr2412.aar) |
| | | | | | | |
| :heavy_check_mark: | 5aeafc0c | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/12/) | 2021-11-26 15:51:05 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-211126-154811-5aeafc-pr2412.zip) |
| :heavy_check_mark: | 5aeafc0c | [#12](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/12/) | 2021-11-26 15:51:45 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-211126-154811-5aeafc-pr2412.zip) |
| :heavy_check_mark: | 5aeafc0c | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/12/) | 2021-11-26 15:54:49 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-211126-154806-5aeafc-pr2412.aar) |
| | | | | | | |
| :heavy_check_mark: | e177defb | [#13](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/13/) | 2021-11-26 15:59:56 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-211126-155821-e177de-pr2412.zip) |
| :heavy_check_mark: | e177defb | [#13](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/13/) | 2021-11-26 16:02:05 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-211126-155816-e177de-pr2412.zip) |
| :heavy_check_mark: | e177defb | [#13](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/13/) | 2021-11-26 16:03:46 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-211126-155816-e177de-pr2412.aar) |
| | | | | | | |
| :heavy_check_mark: | 200daa04 | [#14](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/14/) | 2021-11-26 18:20:13 | ~1 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-211126-181831-200daa-pr2412.zip) |
| :heavy_check_mark: | 200daa04 | [#14](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/14/) | 2021-11-26 18:21:47 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-211126-181831-200daa-pr2412.zip) |
| :heavy_check_mark: | 200daa04 | [#14](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/14/) | 2021-11-26 18:24:15 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-211126-181831-200daa-pr2412.aar) |
| | | | | | | |
| :heavy_check_mark: | 6a203195 | [#15](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/15/) | 2021-11-29 09:14:03 | ~1 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-211129-091223-6a2031-pr2412.zip) |
| :heavy_check_mark: | 6a203195 | [#15](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/15/) | 2021-11-29 09:14:56 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-211129-091223-6a2031-pr2412.zip) |
| :heavy_check_mark: | 6a203195 | [#15](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/15/) | 2021-11-29 09:18:39 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-211129-091223-6a2031-pr2412.aar) |
| | | | | | | |
| :heavy_check_mark: | 8cd179bc | [#16](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/16/) | 2021-11-29 12:11:38 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-211129-120944-8cd179-pr2412.zip) |
| :heavy_check_mark: | 8cd179bc | [#16](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/16/) | 2021-11-29 12:11:40 | ~1 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-211129-120944-8cd179-pr2412.zip) |
| :heavy_check_mark: | 8cd179bc | [#16](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/16/) | 2021-11-29 12:15:10 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-211129-120939-8cd179-pr2412.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | c64b9ba2 | [#17](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/17/) | 2021-11-29 12:13:38 | ~1 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-211129-121209-c64b9b-pr2412.zip) |
| :heavy_check_mark: | c64b9ba2 | [#17](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/17/) | 2021-11-29 12:15:25 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-211129-121210-c64b9b-pr2412.zip) |
| :heavy_check_mark: | c64b9ba2 | [#17](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/17/) | 2021-11-29 12:19:34 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-211129-121512-c64b9b-pr2412.aar) |
| | | | | | | |
| :heavy_check_mark: | 269d3f2c | [#18](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2412/18/) | 2021-11-30 18:40:04 | ~1 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-211130-183822-269d3f-pr2412.zip) |
| :heavy_check_mark: | 269d3f2c | [#18](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2412/18/) | 2021-11-30 18:40:34 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-211130-183822-269d3f-pr2412.zip) |
| :heavy_check_mark: | 269d3f2c | [#18](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2412/18/) | 2021-11-30 18:43:29 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-211130-183817-269d3f-pr2412.aar) |

cammellos commented on protocol/encryption/encryptor.go:
```diff
@@ -592,6 +596,132 @@ func (s *encryptor) EncryptPayload(theirIdentityKey *ecdsa.PublicKey, myIdentity
 	return response, targetedInstallations, nil
 }
 
+func (s *encryptor) getNextKeyID(communityId string) (uint32, error) {
```
This name seems to be a bit generic, I would maybe rename to `getNextHashRatchetKeyID` or similar

cammellos commented on protocol/encryption/encryptor.go:
```diff
@@ -592,6 +596,132 @@ func (s *encryptor) EncryptPayload(theirIdentityKey *ecdsa.PublicKey, myIdentity
+}
+
+// Generates and stores a community key
+func (s *encryptor) GenerateCommunityKey(communityId []byte) (uint32, error) {
```
I would not mention `Community` in general, we can use `GenerateHashRatchetKey` for example

cammellos commented on protocol/encryption/encryptor.go:
```diff
@@ -592,6 +596,132 @@ func (s *encryptor) EncryptPayload(theirIdentityKey *ecdsa.PublicKey, myIdentity
+// Generates and stores a community key
+func (s *encryptor) GenerateCommunityKey(communityId []byte) (uint32, error) {
+
+	//logger := s.logger.With(zap.String("site", "GenerateCommunityKey"))
```
left over

cammellos commented on protocol/encryption/encryptor.go:
```diff
@@ -592,6 +596,132 @@ func (s *encryptor) EncryptPayload(theirIdentityKey *ecdsa.PublicKey, myIdentity
+	if err != nil {
+		return 0, err
+	}
+	communityKeyBytes := crypto.FromECDSA(communityKey) //x509.MarshalECPrivateKey(communityKey)
```
some left over comments I believe

cammellos commented on protocol/encryption/encryptor.go:
```diff
@@ -592,6 +596,132 @@ func (s *encryptor) EncryptPayload(theirIdentityKey *ecdsa.PublicKey, myIdentity
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	logger.Info("encrypting message for communities")
```
maybe `Debug` level?

cammellos commented on protocol/encryption/persistence.go:
```diff
@@ -727,3 +728,153 @@ func (s *sqliteSessionStorage) Load(id []byte) (*dr.State, error) {
+	SeqNo   uint32
+}
+
+// GetCommunityKeyByID retrieves a community key by ID and seqNo. If seqNo = 0, then
```
The comment has been cut short :)

cammellos commented on protocol/encryption/encryptor.go:
```diff
@@ -592,6 +596,132 @@ func (s *encryptor) EncryptPayload(theirIdentityKey *ecdsa.PublicKey, myIdentity
+		return nil, err
+	}
+
+	s.logger.Info("###encryptWithHR", zap.Any("payload", payload))
```
left over comment

cammellos commented on protocol/encryption/encryptor.go:
```diff
@@ -592,6 +596,132 @@ func (s *encryptor) EncryptPayload(theirIdentityKey *ecdsa.PublicKey, myIdentity
+	s.logger.Info("###encryptWithHR", zap.Any("payload", payload))
+	var dbHash []byte
+	if len(hrCache.Hash) == 0 {
+		s.logger.Info("###encryptWithHR no hash")
```
left over comment

cammellos commented on protocol/encryption/encryptor.go:
```diff
@@ -592,6 +596,132 @@ func (s *encryptor) EncryptPayload(theirIdentityKey *ecdsa.PublicKey, myIdentity
+		s.logger.Info("###encryptWithHR no hash")
+		dbHash = []byte(hrCache.Key)
+	} else {
+		s.logger.Info("###encryptWithHR dbhash", zap.Any("hash", hrCache.Hash))
```
left over comment

cammellos commented on protocol/encryption/encryptor.go:
```diff
@@ -592,6 +596,132 @@ func (s *encryptor) EncryptPayload(theirIdentityKey *ecdsa.PublicKey, myIdentity
+		return nil, err
+	}
+	newSeqNo := hrCache.SeqNo + 1
+	s.logger.Info("###encryptWithHR before SaveCommunityKeyHash", zap.Any("new hash", hash.Bytes()),
```
left over comment

cammellos commented on protocol/encryption/encryptor.go:
```diff
@@ -592,6 +596,132 @@ func (s *encryptor) EncryptPayload(theirIdentityKey *ecdsa.PublicKey, myIdentity
+		}
+		// We should not have "holes" in seq numbers,
+		// so a case when hrCache.SeqNo > seqNo shouldn't occur
+		for i := hrCache.SeqNo; i < seqNo; i++ {
```
I would also leave a not here that we should check that `seqNo` is not too distant from `hrCache.SeqNo`, as that can be used as a vector for an attack.
Say `hrCache.SeqNo` is 1 and a user sends us a message with `seqNo` of MAX_UINT64

cammellos commented on protocol/encryption/migrations/sqlite/1632236298_add_communities.up.sql:
```diff
@@ -0,0 +1,16 @@
+CREATE TABLE communities_encryption (
```
I'd not use the `communities` name anywhere in encryption, I would probably use `hash_ratchet` or similar

cammellos commented on protocol/encryption/persistence.go:
```diff
@@ -727,3 +728,153 @@ func (s *sqliteSessionStorage) Load(id []byte) (*dr.State, error) {
+}
+
+// GetCommunityKeyByID retrieves a community key by ID and seqNo. If seqNo = 0, then
+func (s *sqlitePersistence) GetCommunityKeyByID(communityId []byte, keyId uint32, seqNo uint32) (*HRCache, error) {
```
We probably want to cover this with indexes, though can be done separately, but maybe add a comment

cammellos commented on protocol/encryption/persistence.go:
```diff
@@ -727,3 +728,153 @@ func (s *sqliteSessionStorage) Load(id []byte) (*dr.State, error) {
+
+	stmt, err := s.DB.Prepare(`SELECT key_id
+				   FROM communities_encryption
+				     WHERE community_id = ? order by key_id desc limit 1`)
```
Probably should be covered by an index

cammellos commented on protocol/encryption/persistence.go:
```diff
@@ -727,3 +728,153 @@ func (s *sqliteSessionStorage) Load(id []byte) (*dr.State, error) {
+	logger, _ := zap.NewProduction()
+	logger = logger.With(zap.Namespace("Persistence"))
+	err = stmt.QueryRow([]byte(communityId)).Scan(&keyId)
+	// for rows.Next() {
```
leftover

cammellos commented on protocol/encryption/persistence.go:
```diff
@@ -727,3 +728,153 @@ func (s *sqliteSessionStorage) Load(id []byte) (*dr.State, error) {
+	// 	logger.Info("GetCurrentKeyForCommunity row", zap.Any("keyId", keyId))
+
+	// }
+	logger.Info("GetCurrentKeyForCommunity keyCount")
```
maybe debug? or we could remove?

cammellos commented on protocol/encryption/persistence.go:
```diff
@@ -727,3 +728,153 @@ func (s *sqliteSessionStorage) Load(id []byte) (*dr.State, error) {
+	return err
+}
+
+func (s *sqlitePersistence) GetKeyCount() {
```
This looks like a debug function?

cammellos commented on protocol/encryption/persistence.go:
```diff
@@ -727,3 +728,153 @@ func (s *sqliteSessionStorage) Load(id []byte) (*dr.State, error) {
+
+	logger, _ := zap.NewProduction()
+	logger = logger.With(zap.Namespace("Persistence SaveCommunityKey"))
+	logger.Info("Before")
```
left over

cammellos commented on protocol/encryption/protocol_message.proto:
```diff
@@ -54,13 +54,24 @@ message X3DHHeader {
   DHHeader DH_header = 101;
   // Encrypted payload
-  bytes payload = 3;
+  bytes payload = 4;
```
mmh, this looks wrong, it can't be changed,  for backward compatibility, you will have to go out of order with the numbering :)

cammellos commented on protocol/encryption/encryption_test.go:
```diff
@@ -92,6 +94,138 @@ func (s *EncryptionServiceTestSuite) TestGetBundle() {
 }
 
+func (s *EncryptionServiceTestSuite) TestCommunitySend() {
+	// TODO Create a test where several messages
```
Still on TODO?

cammellos reviewed (commented):
Generally looks good, code is very understandable.
Mostly left comments about removing extra logging, and I think we should not mention communities, or at least at the lower levels.
Also a few issues with changing the numbering of protobuf, that's a breaking change and we want to avoid it.

vitvly commented:
Thanks a lot @cammellos ! Per your suggestions:
1. Removed logging remnants and updated some comments.
2. Renamed `Community` to `HashRatchet`
3. Added a delta const to guard against seqNo's arriving too large
4. Created indexes on both tables.

cammellos commented on protocol/encryption/encryption_test.go:
```diff
@@ -69,7 +69,9 @@ func (s *EncryptionServiceTestSuite) initDatabases(config encryptorConfig) {
 
 func (s *EncryptionServiceTestSuite) SetupTest() {
-	s.logger = zap.NewNop()
+	// s.logger = zap.NewNop()
```
Left over probably

Commit:
Disable failing tests

Fix lint
